### DISCUSSION TITLE
message.success is not a function

### DISCUSSION THREAD

**Discussion by 2010363329** [ant-design/ant-design#46237] [Q&A] [answered]
看不太懂 跟官网一模一样的用法：https://ant-design.antgroup.com/components/app-cn#%E4%BB%A3%E7%A0%81%E6%BC%94%E7%A4%BA

```
successMsg(msg);

const successMsg = (msg: string) => {
  message.success(msg).then();
};
```
这么写就报错了 app也引入了 我很好奇 message.success难道不是一个方法吗

**Comment by Wxh16144** [comment]
你可能没有注意到后半段说明🤔
<img width="1417" alt="image" src="https://github.com/ant-design/ant-design/assets/32004925/374b268a-62f1-4824-8cbd-f6deaaf4d5a3">

**Reply by 2010363329** [reply]
<img width="439" alt="image" src="https://github.com/ant-design/ant-design/assets/76477581/53c8ba31-58e0-456f-b854-41aa74a9990a">
也没错 都看了

**Reply by Wxh16144** [reply]
截图看不出问题呀，你调用 `App.useApp()` hook 的地方应该出现在 `<App />` 嵌套的组件中

**Reply by 2010363329** [reply]
确实 厉害 没有认识到层次的问题 改过来了 已经
之前的：
<img width="625" alt="image" src="https://github.com/ant-design/ant-design/assets/76477581/8880aa56-c181-48d0-b4da-0e1d0f2c65da">
之后的：
<img width="532" alt="image" src="https://github.com/ant-design/ant-design/assets/76477581/101e8aba-9e7d-4c65-8ad3-5299f894a991">
因为是第一次用React+ant（新手）我想知道这两种代码层次有什么区别吗 以及这两种写法的优劣势是什么 还是说没什么区别 因为自己项目很多已经是按照图一写得 后面要是改很多文件都得改 能不能详细分析一下 或者有相关的链接可以学习的

**Reply by Wxh16144** [reply]
> 这两种代码层次有什么区别吗？

我口述可能比较含糊，直接看 React 文档吧：https://react.dev/reference/react/useContext#my-component-doesnt-see-the-value-from-my-provider

**Reply by 2010363329** [reply]
三克油 我看看

**Comment by zombieJ** [comment, marked as answer]
没代码，盲猜你在使用 App 的代码同层调用了 `App.useApp`。`useApp` 必须在子元素下调用，不能同级调用。

**Reply by 2010363329** [reply]
确实 厉害 没有认识到层次的问题 改过来了 已经
之前的：
<img width="625" alt="image" src="https://github.com/ant-design/ant-design/assets/76477581/8880aa56-c181-48d0-b4da-0e1d0f2c65da">
之后的：
<img width="532" alt="image" src="https://github.com/ant-design/ant-design/assets/76477581/101e8aba-9e7d-4c65-8ad3-5299f894a991">
因为是第一次用React+ant（新手）我想知道这两种代码层次有什么区别吗 以及这两种写法的优劣势是什么 还是说没什么区别 因为自己项目很多已经是按照图一写得 后面要是改很多文件都得改 能不能详细分析一下 或者有相关的链接可以学习的
